File: ecommerce-api-legacy/src/middlewares/errorHandler.js
'use strict';

class BusinessError extends Error {
    constructor(message, statusCode = 400) {
        super(message);
        this.statusCode = statusCode;
        this.name = this.constructor.name;
    }
}

class NotFoundError extends BusinessError {
    constructor(message = 'Recurso não encontrado') { super(message, 404); }
}

class UnauthorizedError extends BusinessError {
    constructor(message = 'Não autorizado') { super(message, 401); }
}

class ValidationError extends BusinessError {
    constructor(message = 'Dados inválidos') { super(message, 400); }
}

function errorHandler(err, _req, res, _next) {
    if (err instanceof BusinessError) {
        return res.status(err.statusCode).json({ error: err.message });
    }
    console.error('[ERROR]', err.stack || err.message);
    return res.status(500).json({ error: 'Erro interno do servidor' });
}

module.exports = { BusinessError, NotFoundError, UnauthorizedError, ValidationError, errorHandler };

File: task-manager-api/middlewares/error_handler.py
import logging

from flask import jsonify


logger = logging.getLogger(__name__)


class BusinessError(Exception):
    status_code = 400

    def __init__(self, message: str, status_code: int | None = None):
        super().__init__(message)
        if status_code is not None:
            self.status_code = status_code


class ValidationError(BusinessError):
    status_code = 400


class NotFoundError(BusinessError):
    status_code = 404


class ConflictError(BusinessError):
    status_code = 409


class UnauthorizedError(BusinessError):
    status_code = 401


class ForbiddenError(BusinessError):
    status_code = 403


def register_error_handlers(app) -> None:
    @app.errorhandler(BusinessError)
    def _handle_business(error: BusinessError):
        return jsonify({"error": str(error)}), error.status_code

    @app.errorhandler(404)
    def _handle_not_found(_error):
        return jsonify({"error": "Recurso não encontrado"}), 404

    @app.errorhandler(Exception)
    def _handle_generic(error: Exception):
        logger.exception("Erro não tratado: %s", error)
        return jsonify({"error": "Erro interno do servidor"}), 500
